File: scala/src/test/scala/bbc/dojonorth/lyrics/MusixMatchSongMatcherSpec.scala
package bbc.dojonorth.lyrics

import org.scalatest.{Matchers, FlatSpec}
import scala.language.postfixOps
import scala.collection.immutable.Nil

class MusixMatchSongMatcherSpec extends FlatSpec with Matchers {
  val fixturesDir = "./src/test/resources"
  val testFile = s"$fixturesDir/mxm_sample.txt"

  val musixMatchSongMatcher = new MusixMatchSongMatcher

  "A song matcher data loader" should "ignore comment lines" in {
    val lines = musixMatchSongMatcher.loadDataFile(testFile)
    val commentedLines = lines filter { line => line.startsWith("#") } toList

    commentedLines should be (Nil)
  }

  it should "give an iterator of strings containing <SEP>" in {
    val lines = musixMatchSongMatcher.loadDataFile(testFile)

    lines foreach { line =>
      line should include ("<SEP>")
    }
  }

  val exampleTrack = MSDTrack("TRMMMKD128F425225D",4418550,"Karkkiautomaatti", "Tanssi vaan")

  "A mxm line parser" should "parse lines beginning with TR into song descriptors" in {

    val exampleLine = "TRMMMKD128F425225D<SEP>Karkkiautomaatti<SEP>Tanssi vaan<SEP>4418550<SEP>Karkkiautomaatti<SEP>Tanssi vaan"

    val parsed = musixMatchSongMatcher.parseMsdTrackLine(exampleLine)

    parsed should be (exampleTrack)
  }

  "A song metadata loader" should "parse all lines into track data" in {

    val metadata = musixMatchSongMatcher.loadSongMetaData(testFile)

    metadata.keys.size should be (4)
    metadata.values should contain (exampleTrack)
  }
}

File: javascript/test/lyrics/MusixMatchSongMatcherSpec.js
var assert = require('assert');
var _ = require('lodash');
var MusixMatchSongMatcher = require('../../lib/lyrics/MusixMatchSongMatcher');

describe('MusixMatchSongMatcher', function () {
    var musixMatchSongMatcher, fixturesDir, testFile;
    before(function () {
        musixMatchSongMatcher = new MusixMatchSongMatcher();
        fixturesDir = './test/resources';
        testFile = fixturesDir + '/mxm_sample.txt';
    });

    describe('#loadDataFile()', function () {
        it('should ignore comment lines', function () {
            var lines = musixMatchSongMatcher.loadDataFile(testFile);
            var commentedLines = _.filter(lines, function(line) {
                return line.indexOf('#') === 0;
            });

            assert.equal(commentedLines.length, 0);
        });

        it('should give an iterator of strings containing <SEP>', function () {
            var lines = musixMatchSongMatcher.loadDataFile(testFile);
            _.each(lines, function (line) {
                assert.notEqual(-1, line.indexOf('<SEP>'));
            });
        });
    });

    describe('#parseMsdTrackLine()', function () {
        var exampleTrack = {trackId: "TRMMMKD128F425225D", mxmId: 4418550, artist: "Karkkiautomaatti", title: "Tanssi vaan"};

        it('should parse lines beginning with TR into song descriptors', function() {
            var exampleLine = 'TRMMMKD128F425225D<SEP>Karkkiautomaatti<SEP>Tanssi vaan<SEP>4418550<SEP>Karkkiautomaatti<SEP>Tanssi vaan';
            var parsed = musixMatchSongMatcher.parseMsdTrackLine(exampleLine);

            // this is a hacky as hell way of JS object comparison,
            // it may fail depending on internal JSON implementation.
            assert.equal(JSON.stringify(exampleTrack), JSON.stringify(parsed));
        });
    });

    describe('#loadSongMetaData()', function () {
        it('should parse all lines into track data', function() {
            var metadata = musixMatchSongMatcher.loadSongMetaData(testFile);
            assert.equal(4, Object.keys(metadata).length);
        });
    });
});